src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3377,9 +3448,56 @@ private void VisitTupleExpression(BoundTupleExpression node)
             var arguments = node.Arguments;
             ImmutableArray<TypeSymbolWithAnnotations> elementTypes = arguments.SelectAsArray((a, w) => w.VisitRvalueWithResult(a), this);
             var tupleOpt = (TupleTypeSymbol)node.Type;
-            _resultType = (tupleOpt is null) ?
-                default :
-                TypeSymbolWithAnnotations.Create(tupleOpt.WithElementTypes(elementTypes), NullableAnnotation.NotNullable);
+            if (tupleOpt is null)
+            {
+                _resultType = default;
+            }
+            else
+            {
+                int slot = GetOrCreateObjectCreationPlaceholderSlot(node);
+                if (slot > 0)
+                {
+                    this.State[slot] = NullableAnnotation.NotNullable;
+                    TrackNullableStateOfTupleElements(slot, tupleOpt, arguments, elementTypes, useRestField: false);
+                }
+                _resultType = TypeSymbolWithAnnotations.Create(tupleOpt.WithElementTypes(elementTypes), NullableAnnotation.NotNullable);
+            }
+        }
+
+        private void TrackNullableStateOfTupleElements(
```

Comment:
>TrackNullableStateOfTupleElements [](start = 21, length = 33)

A comment on `useRestField` would be helpful. For tuple literals, we can have N elements/values in a flat list. But for tuples constructed from `new ValueTuple(...)` the 8th element will contain nested state to track.
Btw, should we recognize `ValueTuple.Create(...)` too? (not essential for this PR)
Not related to this PR: should we recognize the `Nullable<T>` constructor too?

Reply:
Logged [https://github.com/dotnet/roslyn/issues/32625](https://github.com/dotnet/roslyn/issues/32625) for `ValueTuple.Create`. And logged [https://github.com/dotnet/roslyn/issues/32626](https://github.com/dotnet/roslyn/issues/32626) for `Nullable<T>(T value)`. Thanks.

---
In reply to: [249165883](https://github.com/dotnet/roslyn/pull/32612#discussion_r249165883) [](ancestors = 249165883)